quarkchain/cluster/shard_state.py

```diff
@@ -1523,3 +1508,49 @@ def _get_evm_state_from_height(self, height: Optional[int]) -> Optional[EvmState
             Logger.error("Failed to get block at height {}".format(height))
             return None
         return self._get_evm_state_for_new_block(block)
+
+    def __get_coinbase_addresses_until_block(
+        self, header_hash: bytes, length: int
+    ) -> List[bytes]:
+        """Get coinbase addresses up until block of given hash within the window."""
+        curr_block = self.db.get_minor_block_by_hash(header_hash)
+        if not curr_block:
+            raise ValueError("curr block not found: hash %d" % header_hash)
+        header = curr_block.header
+        height = header.height
+        prev_hash = header.hash_prev_minor_block
+        if prev_hash in self.coinbase_addr_cache:  # mem cache hit
+            _, addrs = self.coinbase_addr_cache[prev_hash]
+            addrs = addrs.copy()
+            if len(addrs) == length:
+                addrs.popleft()
+            addrs.append(header.coinbase_address.recipient)
+        else:  # miss, iterating DB
+            addrs = deque()
+            for _ in range(length):
+                addrs.appendleft(header.coinbase_address.recipient)
+                header = self.db.get_minor_block_header_by_hash(
+                    header.hash_prev_minor_block
+                )
+                if not header:
+                    break
+        self.coinbase_addr_cache[header_hash] = (height, addrs)
+        # in case cached too much, clean up
+        if len(self.coinbase_addr_cache) > 128:  # size around 640KB if window size 256
+            self.coinbase_addr_cache = {
+                k: (h, addrs)
+                for k, (h, addrs) in self.coinbase_addr_cache.items()
+                if h > height - 16  # keep most recent ones
+            }
+        return list(addrs)
+
+    @functools.lru_cache(maxsize=16)
+    def _get_posw_coinbase_blockcnt(self, header_hash: bytes) -> Dict[bytes, int]:
```

Comment:
a more generalized method `__get_coinbase_addresses_until_block` is provided in case similar functionalities are needed (which doesn't include any PoSW-related param). this method simply serves as the wrapper over that.